akka-projection-slick/src/main/scala/akka/projection/slick/SlickProjection.scala

```diff
@@ -23,19 +23,44 @@ object SlickProjection {
   def exactlyOnce[Offset, Envelope, P <: JdbcProfile: ClassTag](
       projectionId: ProjectionId,
       sourceProvider: SourceProvider[Offset, Envelope],
-      databaseConfig: DatabaseConfig[P])(eventHandler: Envelope => DBIO[Done]): Projection[Envelope] =
+      databaseConfig: DatabaseConfig[P],
+      eventHandler: SlickEventHandler[Envelope]): Projection[Envelope] =
     new SlickProjectionImpl(projectionId, sourceProvider, databaseConfig, SlickProjectionImpl.ExactlyOnce, eventHandler)
 
   def atLeastOnce[Offset, Envelope, P <: JdbcProfile: ClassTag](
       projectionId: ProjectionId,
       sourceProvider: SourceProvider[Offset, Envelope],
       databaseConfig: DatabaseConfig[P],
       saveOffsetAfterEnvelopes: Int,
-      saveOffsetAfterDuration: FiniteDuration)(eventHandler: Envelope => DBIO[Done]): Projection[Envelope] =
+      saveOffsetAfterDuration: FiniteDuration,
+      eventHandler: SlickEventHandler[Envelope]): Projection[Envelope] =
     new SlickProjectionImpl(
       projectionId,
       sourceProvider,
       databaseConfig,
       SlickProjectionImpl.AtLeastOnce(saveOffsetAfterEnvelopes, saveOffsetAfterDuration),
       eventHandler)
+
+}
+
+trait SlickEventHandler[Envelope] {
+  def handleEvent(envelope: Envelope): DBIO[Done]
+  def onFailure(envelope: Envelope, throwable: Throwable): RecoverStrategy = {
+    val _ = envelope // need it otherwise compiler says no
+    val _ = throwable
+    RecoverStrategy.fail
+  }
+}
+
+sealed trait RecoverStrategy
+case object Fail extends RecoverStrategy
+case object Skip extends RecoverStrategy
+final case class RetryAndFail(retries: Int, delay: FiniteDuration) extends RecoverStrategy
+final case class RetryAndSkip(retries: Int, delay: FiniteDuration) extends RecoverStrategy
+
+object RecoverStrategy {
+  def fail: RecoverStrategy = Fail
+  def skip: RecoverStrategy = Skip
+  def retryAndFail(retries: Int, delay: FiniteDuration): RecoverStrategy = RetryAndFail(retries, delay)
+  def retryAndSkip(retries: Int, delay: FiniteDuration): RecoverStrategy = RetryAndSkip(retries, delay)
```

Comment:
`fail` and `retryAndFail` may not be necessary if we always wrap the source with a `RestartSource`. 

The Pros is that the retry can be much faster without stopping the whole stream and restarting
The Cons being that after retrying and failing, the projection will start over after the backoff period. And retry again and again. That may be confusing for users.

Reply:
We also may want to make this part of the core api and reuse in the other projecitons

Reply:
Would the retry include handling transient reconnection errors to the underlying database impl. too?

Reply:
> Would the retry include handling transient reconnection errors to the underlying database impl. too?

That's up to the user. The `onFailure` receive the 'offending' event and the exception. User could then decide if it's a retry, how many times, which delay, ...